pkg/trace/traceutil/truncate.go

```diff
@@ -5,19 +5,24 @@
 
 package traceutil
 
+import "unicode/utf8"
+
 // TruncateUTF8 truncates the given string to make sure it uses less than limit bytes.
 // If the last character is an utf8 character that would be splitten, it removes it
 // entirely to make sure the resulting string is not broken.
 func TruncateUTF8(s string, limit int) string {
 	if len(s) <= limit {
 		return s
 	}
-	var lastValidIndex int
-	for i := range s {
-		if i > limit {
-			return s[:lastValidIndex]
-		}
-		lastValidIndex = i
+	s = s[:limit]
+	// The max length of a valid code point is 4 bytes, therefore if we see all valid
+	// code points in the last 4 bytes we know we have a fully valid utf-8 string
+	// If not we can truncate one byte at a time until the end of the string is valid utf-8
+	for !(len(s) >= 4 && utf8.Valid([]byte(s[len(s)-4:]))) &&
+		!(len(s) >= 3 && utf8.Valid([]byte(s[len(s)-3:]))) &&
+		!(len(s) >= 2 && utf8.Valid([]byte(s[len(s)-2:]))) &&
+		!(len(s) >= 1 && utf8.Valid([]byte(s[len(s)-1:]))) && len(s) >= 1 {
+		s = s[:len(s)-1]
 	}
```

Comment:
It feels a bit crammed, although it's quite clear what it's doing (because of the comment). I would suggest to unpack the for-loop condition into something like this:

```suggestion
	for len(s) >= 1 {
		if len(s) >= 4 && utf8.Valid([]byte(s[len(s)-4:])) {
			break
		}
		if len(s) >= 3 && utf8.Valid([]byte(s[len(s)-3:])) {
			break
		}
		if len(s) >= 2 && utf8.Valid([]byte(s[len(s)-2:])) {
			break
		}
		if len(s) >= 1 && utf8.Valid([]byte(s[len(s)-1:])) {
			break
		}
		s = s[:len(s)-1]
	}
```

I tested it and it passes the tests, so it seems to work. Also, I've been fuzzing it for about 10 minutes, with no errors.